rfcs/0124-worker-manager.md

```diff
@@ -0,0 +1,271 @@
+  concept
+- Scaling Ratio: the ratio of pending tasks to running workers which the worker
+  manager will try to maintain.  Used for hard-to-provision systems or
+  heavy-burst loads so that workers are not started before demand is really
+  there
+- Worker: this is the name for a system which is running a Taskcluster worker
+  program.  A worker is roughly equivalent to an EC2 instance running a worker
+  program.
+- Worker Program: This is a program which pulls tasks off the task Queue and
+  runs them.  This term refers to the codebase, like generic-worker
+- Worker Image: This is a snapshot of a specific machine before it is booted.
+  Used with broad equivalence to the AMI (Amazon Machine Image) term
+- Worker Id: The cloud specified resource identifier for a worker.  In AWS, it
+  is the instanceId field, example: i-abcdef0123456789
+- Worker Group: The cloud specified resource identifier for the region.  In AWS,
+  it is the region's programmatic name, example: us-east-1
+- Worker Configuration: (new term) this is an entry in the worker manager's
+  API.  It will have an identifier, and that identifier will only be used
+  within the worker manager
+- Worker Type: (new term) name of a pool of workers as used by the Queue.
+  Exclusively an identifier.  May have more than one worker Type for a given
+  worker configuration.  
+
+## Architecture Overview
+The worker manager will be comprised of a Taskcluster API, a service to listen
+to Queue messages, a set of providers, a set of bidding strategies, and
+credential generation.  It will be implemented in Node.js using standard
+Taskcluster services libraries.
+
+There is no intent to create a full Taskcluster service for each pluggable
+component.  The interchange between pluggable components and the main worker
+manager will be through Javascript method calling.  It is the discretion of the
+component's author whether the full component runs in the Worker Manager's
+process or if it functions as a bridge to an external service.  These
+components are intended to be developed as part of the worker manager codebase.
+
+## API
+The worker manager will expose APIs to control workers.  These include terminating
+instances and listing running workers.  There will also be APIs which operate
+on whole worker configurations and worker types.  These APIs will be standard
+Taskcluster APIs and will use scopes which are categorized by worker configuration
+and worker type.
+
+### Worker State
+Worker manager will also provide worker state APIs.  These APIs will show
+information like the last task runs for a given worker type, worker
+configuration or worker.  The data backing this information will be received
+from the Queue over RabbitMQ.  The worker manager will track outcomes.  The
+only information from the Queue stored in this service will be the task run
+outcome, task id and run id.  All other information will be available through
+the Queue.
+
+### Worker Configuration
```

Comment:
This section is confusing to me. I feel like maybe it would benefit from a small example/diagram/codeblock. Specifically the _general_ structure of worker configuration. You can state explicitly that it is for example purposes and is not set in stone yet if you'd like.

I got confused when we got to the `disabled` part and I was trying to visualize what the tree of config sections looked like.

Reply:
I've reworked this.  I agree it was underspecified.  Let me know what you think of the new section for worker configuration.